File: 190ReverseBits.cpp
/*
Reverse bits of a given 32 bits unsigned integer.

Example:

Input: 43261596
Output: 964176192
Explanation: 43261596 represented in binary as 00000010100101000001111010011100, 
             return 964176192 represented in binary as 00111001011110000010100101000000.

Follow up:
If this function is called many times, how would you optimize it?
*/


class Solution{
public:
	uint32_t reverseBits(uint32_t n){
		if(n == 0) return 0;
		vector<int> bits;
		for(int i =0; i < n; ++i){
			bits.push_back(n%2);
			n >>= 1;
		}
		uint32_t m =0;
		for(int j = 0; j<32; ++j){
			m = 2*m + bits[i];
		}
		return m;
	}
}


//通过右移n依次获取n的末位，设置为m的末尾，左移m实现颠倒

	/*
	uint32_t reverseBit(uint32_t n){
		uint32_t m = 0;
		while(n != 0){
			m <<= 1;
			m = m | (n & 1);
			n >>= 1;
		}
		return m;
	}
	
	*/

File: 190ReverseBits.py
class Solution:
    # @param n, an integer
    # @return an integer
    '''
	先将输入转换成2进制字符串，再翻转并扩充到32位，再将此32位的二进制转为无符号整数即可。利用Python的bin()函数很方便。
    '''
    def reverseBits(self, n):
    	res = bin(n)[:1:-1]
    	return int(res + '0'*(32-len(res)), 2)

   '''
   按位处理，将输入n的二进制表示从低位到高位的值依次取出，逆序排列得到翻转后的值。这里更新res的时候，用纯位操作会比用加法要快的多。
   '''

   def reverseBits(self, n):
    	"""
        :type n: int
        :rtype: int
        """
        res = 0
        for i in xrange(32):
            res <<= 1
            res |= ((n >> i) & 1)
        return res


    